### ISSUE TITLE
Side-effects of settings defaultZoneName

### ISSUE THREAD

**Issue by evenicoulddoit** [moment/luxon#378]
I came across this whilst trying to mock dates within tests, and found it rather un-intuitive, so thought I'd raise after having a quick check for any mentions previous.

As part of my tests, I wanted to change the timezone, so I changed the value of `Settings.defaultZoneName` in order to accomplish this. To unpatch the setting, I captured its previous value and re-instated, e.g.:

```js
someTestCase(() => {
  let original;

  beforeEach(() => {
    Settings.defaultZoneName = 'Asia/Singapore';
  });

  afterEach(() => {
    Settings.defaultZoneName = original;
  });
}
```

However, after debugging and issue I found that this is not in fact restoring the original state:

* `Settings.defaultZoneName`is [really just a proxy to `Settings.defaultZone`](https://github.com/moment/luxon/blob/master/src/settings.js#L52)
* [`Settings.defaultZone` is initially a `LocalZone` instance]([https://github.com/moment/luxon/blob/master/src/settings.js#L61]) because [the private `defaultZone` variable is set to null](https://github.com/moment/luxon/blob/master/src/settings.js#L7)
* As a result, when you set `Settings.defaultZoneName` to be *anything*, [it is normalized to produce a non-`LocalZone`](https://github.com/moment/luxon/blob/master/src/impl/zoneUtil.js#L13:17)

The end result is that you end up with this, slightly confusing (at least in my mind) behaviour:

```JS
luxon.Settings.defaultZone
// OUT: LocalZone {}

// Set the defaultZoneName to equal its current value
luxon.Settings.defaultZoneName = luxon.Settings.defaultZoneName

luxon.Settings.defaultZone
// OUT: IANAZone {zoneName: "Asia/Singapore", valid: true}
```

My initial thought would be to have the `defaultZoneName` return `null` when the `defaultZone` is a `LocalZone` instance, but I'm not well enough versed in the Luxon codebase to understand the ramifications of that change.

**Comment by icambron** [comment]
Yeah, this probably needs rethought. In general, I've avoided letting users manipulate zone settings at the zone object level and made everyone do it at the name level, but I'm considering changing that.

**Comment by icambron** [comment]
I think I will make the breaking change you suggest, which is to make `defaultZoneName` return `null` by default. I think that should work fine for most cases, because internally Luxon is always looking for `defaultZone` itself, never the name.

**Comment by icambron** [comment]
I'm doing this in 2.0. It works like this:

 * `LocalZone` is now called `SystemZone`. You can get at it directly with "system" (as in `dt.setZone("system")`
 * The getter and setter for `Settings.defaultZoneName` just manage a simple string value
 * The default value of that string is "system"
 * The getter for `Settings.defaultZone` gets an instance of the zone matching the string (I may later cache this result if there's a performance benefit to it, but conceptually it is still just this)

You can see it in 5dca0eb

**Comment by evenicoulddoit** [comment]
Looks good! Might be nice to use constants for clarity, also this `helpers` module looks like it could be useful for testing 3rd party libs, is it included in the build?

**Comment by icambron** [comment]
On the constants, I keep thinking I want to do that and then realize that people can just use `SystemZone.instance` if they want that.

On the `helpers` module, it's just test infrastructure. I can see how it's broadly useful, but I also think it's easy to replicate, so I'm hesitant to add it because I want to keep the surface area I support as small as possible.

**Comment by evenicoulddoit** [comment]
Fair enough on both points! 🙂

**Comment by GillesDebunne** [comment]
What about removing `Settings.defaultZoneName` altogether? No more proxy, but a new `Settings.defaultZone` *setter* instead.

Just like `DateTime.setZone()` (and actually all the methods accepting a zone, which all end up calling `normalizeZone`), the `Settings.defaultZone` setter would accept `null`, `undefined`, _'default'_, _'system'_, _'Asia/Tokyo'_ as well as Zone instances.

The problematic `luxon.Settings.defaultZone = luxon.Settings.defaultZone` line above would no longer be an issue.

The `defaultZone` getter does not always returns what was passed to the setter, since it returns the normalized Zone, but that's consistent with all the other usages of a zone parameter.

**Comment by icambron** [comment]
@GillesDebunne I think that makes sense: it certainly cleans up the interface and it makes any worries about caching go away. One nit: it can't accept "default", right?

**Comment by GillesDebunne** [comment]
Indeed. Since it feels more consistent to accept `undefined` and `null`, and we need to define a behavior in that case, I think it makes sense to accept the same set of zone options. The "default default" being SystemZone. That's actually what `defaultZoneName = 'default'` does today.

I have this PR ready on my 2.0 branch, but I saw that you rebased `2.0-take-2` on `master`, however it does not include the changes I submitted that got merged in the last few days.

**Comment by icambron** [comment]
Ooops. Where can I find those commits? I'll add them back

**Comment by icambron** [comment]
Nm, I can just snag them from the PR

**Comment by icambron** [comment]
Fixed in 2.0
